### PR TITLE
Stop Dynamo from importing optional libraries at import time

### LINKED ISSUE #142489: Stop importing arbitrary third-party libraries in Dynamo
The problem is that arbitrary third-party libraries can have side effects on import. Instead, we should check if the libraries have already been imported. Example of what we should do: https://github.com/pytorch/pytorch/blob/34127fc6880e25e71c62daf8fc8f909a4efddd2d/torch/_dynamo/variables/dicts.py#L712

Also, part of this issue is figuring out how to add a lint rule (against `import <third_party_library>` in torch/_dynamo) so that this doesn't happen again.

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #186044

Dynamo still had a few import-time paths that loaded optional third-party
packages. The most visible one was Triton: importing torch._dynamo could load
Triton through torch._dynamo.utils.has_triton_package() and through Inductor
runtime hint setup reached from guard imports. Other code paths imported or
looked up optional packages such as einops, optree, torchrec, and _pytest
instead of only acting on modules that were already present.

Fix this by changing Dynamo's optional-library integrations to use sys.modules
when the library must already be loaded, by deferring optree pytree polyfill
registration until optree initializes, and by lazily resolving Triton-specific
AttrsDescriptor handling in Inductor hints. The import linter now has an
import-time rule for torch/_dynamo that rejects top-level imports of optional
third-party packages and top-level has_triton_package() calls, which covers the
root cause that allowed this to regress.

This keeps local runtime imports allowed where they are intentional, instead of
banning all optional library use in Dynamo. A broader dynamic-import detector
would catch more patterns, but it would also be noisy and harder to maintain;
the syntactic rule covers the observed root-cause patterns directly.

Fixes #142489
Generated by my agent

Test Plan:
- python -m unittest tools.test.test_import_linter
- python tools/linter/adapters/import_linter.py -- $(rg --files torch/_dynamo -g '*.py')
- python - <<'PY' ... import torch; before = set(sys.modules); import torch._dynamo ... PY
  Verified no new modules rooted at einops, optree, torchrec, transformers,
  diffusers, pandas, sklearn, or triton were loaded.
- python - <<'PY' ... import torch._inductor.runtime.hints as hints ... PY
  Verified importing hints does not load Triton and fallback AttrsDescriptor
  creation still works.
- python - <<'PY' import triton.language as tl; from torch._dynamo.utils import is_safe_constant; print(is_safe_constant(tl.float32)) PY
- python - <<'PY' ... torch.compile(f, backend="eager") ... PY
- lintrunner -a tools/linter/adapters/import_linter.py tools/test/test_import_linter.py torch/_dynamo/decorators.py torch/_dynamo/variables/user_defined.py torch/_dynamo/utils.py torch/_dynamo/logging.py torch/_dynamo/polyfills/loader.py torch/_dynamo/polyfills/pytree.py torch/_inductor/runtime/hints.py
- git diff --check

Benchmark Results:
Fresh-process microbenchmark timing only `import torch._dynamo` after `import torch`,
5 runs each. Baseline main at 5011b03ed47 used a temporary worktree with built
library symlinks from this checkout.

- main: raw seconds [1.480360301968176, 0.68467306898674, 0.6751202560262755, 0.7102447140496224, 0.6831374939647503], mean 0.8467071669991129
- fixed branch: raw seconds [0.6633166190003976, 0.6530681599979289, 0.647358551970683, 0.6154148380155675, 0.6140483449562453], mean 0.6386413027881644

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says the inductor_cpp_wrapper failure was caused by tl.float32 no longer being wrapped as a constant after deferring Triton imports. I updated Dynamo to recognize already-loaded Triton dtype objects as identity-guarded constants, kept optional DTensor/FSDP paths lazy, and added coverage for the DTensor rule-map cache behavior.

**Comment by jansel** [comment]
My agent says the broad CI failures were caused by Inductor importing lowerings before the Python-defined symm_mem low-contention schemas were registered, so torch.ops.symm_mem._low_contention_all_gather was missing during test imports. I updated Inductor symm_mem lowering registration to import the internal schema owner before probing torch.ops.symm_mem, preserving the low-contention lowerings while fixing the import-order failure.

**Comment by jansel** [comment]
My agent says the remaining distributed/Inductor CI failures came from tests relying on registrations that Dynamo no longer performs as import-time side effects. I updated those tests to import the FSDP schema owner for torch.ops.fsdp.copy_ and the fake process-group module before using backend="fake", preserving the lazy production import path while making the test dependencies explicit.

**Comment by jansel** [comment]
My agent says the latest distributed CI failures were another test dependency on the old accidental fake process-group registration side effect. I updated test_init_process_group_for_all_backends to import the fake process-group test module explicitly and use FakeStore for Backend.FAKE, keeping Dynamo import-time behavior lazy while making the distributed test own the backend it enumerates.

**Review by XuehaiPan — APPROVED** [review]

**Comment by jansel** [comment]
My agent says I rebased the PR and fixed the latest related CI issues: the import linter now skips non-Python files like torch/_dynamo/CLAUDE.md, mark_unbacked no longer assumes torch.distributed.tensor was imported, and the affected tests now explicitly register the FSDP/fake process-group owners they rely on while guarding platforms without c10d. I also reran the targeted failing tests locally.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I addressed the latest review feedback by caching the safe-constant type set after Triton dtype resolution, making lazy module init run immediately for already-loaded modules, and moving the DTensor rule-map regression coverage into a fresh subprocess so it no longer skips when the test file has already loaded DTensor. I also added linter coverage for class-body optional imports and import-time default-argument has_triton_package() calls, then reran the targeted tests and linters.

**Comment by jansel** [comment]
My agent says I pushed a follow-up for the AOTInductor logging failure. The logging test helper now counts and patches only handlers created by torch._logging, using the same _is_torch_handler ownership marker that torch logging uses for cleanup, so extra runner/test handlers on registered PT2 loggers do not trip the PT2 handler-count assertion.

I intentionally left the benchmark expected CSVs unchanged. The mobilenet/mnasnet benchmark failures are eager_two_runs_differ before Dynamo or Inductor runs (compilation time 0, 0 graphs), so baselining those rows here would mask unrelated eager nondeterminism rather than fix this PR. Local checks run: lintrunner -a, python test/dynamo/test_logging.py LoggingTests.test_skip_reason_logging -v, a handler-watcher smoke test with extra non-torch handlers, and git diff --check.

**Comment by jansel** [comment]
My agent says I updated the pr_time_benchmarks DTensor dispatch expected results from the CI-generated expected file. Only custom_handler crossed the 10% guardrail, but the checker asked to refresh all significantly changed DTensor dispatch rows, so I updated the six DTensor rows it printed. I also reran lintrunner, import-linter/unit tests, DTensor trace-rule tests, the optional-import smoke test, and a CUDA Inductor torch.put smoke test. The prior ROCm test_put failure was a CUDA driver error 401, so this push will rerun it on fresh CI.

**Comment by jansel** [comment]
My agent says I rebased the PR onto current main to pick up the ROCm MI350 runner-label workflow changes, resolved the DTensor benchmark CSV conflict by preserving the prior CI-generated DTensor expected values, and addressed fresh review feedback for AttrsDescriptorWrapper fallback defaults plus lambda default-argument import-time linting. Local checks rerun: lintrunner -a, import-linter unit tests, Dynamo import-linter sweep, trace-rule tests, import-time smoke test, and the new AttrsDescriptorWrapper no-Triton fallback regression.

**Comment by jansel** [comment]
My agent says I rebased onto current main, preserving #190649 PyTreeSpec equality/hash behavior while retaining lazy optree loading. The update also adds an end-to-end Dynamo import-hygiene regression, preserves dynamic polyfill rules across the DTensor rule-map transition, completes optional import lint coverage, and narrows symmetric-memory schema-owner error handling. Local validation passed: lintrunner -a; the upstream pytree/hash suites (111 passed, 14 skipped); and targeted import-linter, import, DTensor, FSDP, distributed, and logging tests.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says the late optree review finding was valid: `_load_pytree_polyfill_module()` now clears `guards.uninteresting_files()` after extending `POLYFILLED_MODULES`, with a regression that primes the cache before a synthetic late polyfill load. I also corrected the lazy-init docstring. The existing `test_triton_kernel_triton_dtype` already covers `tl.float32` through `torch.compile(..., fullgraph=True)` across eager, AOT eager, and Inductor. The prior CI failures were unrelated: TIMM could not load eight Hugging Face models in offline mode, and two ROCm shards timed out before pytest created test state; fresh CI will rerun on this update.

**Comment by jansel** [comment]
My agent says the CUDA distributed failure was caused by the TorchComms mixed-backend test relying on Dynamo's former accidental import of the fake process-group registration. I rebased onto current main and made the test register cpu:fake explicitly; I also tightened the import linter's alias handling and removed broad test-only ImportError catches found during review. Targeted distributed/import tests and lintrunner -a pass; the exact TorchComms test skips locally because TorchComms is not installed.

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
